mitodl/open-discussions
# Reindex posts and comments from DB

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes #1700 
Closes #1662 

#### What's this PR do?

Final step of this epic, refactors the search indexing to run off the DB rather than reddit. My main goals here were to ensure zero or minimal impact to non-search parts of the application.

Specifically it includes these improvements:

- Search serializers for posts and comments are now their own implementation and updated to serialize directly from `Post` and `Comment` since they do not need all fields the API serializers use anyway. This avoids additional overhead of fetching data for fields we don't write to ES.
- Reindexes off a prefetched iterator to keep DB trips to a minimum
- Moves comment reindexing to its own task since it can now be done independently of the submission/post

#### How should this be manually tested?

- Run `./manage.py recreate_index` with reddit shut off, it should not error
- Verify you can still search for items and the results are as expected
- Check the record count for `Post` and `Comment` match the corresponding document counts in ES. E.g.:

```bash
curl --header "Content-Type: application/json" \
  --request POST -s \
  --data '{"query": {"bool": {"must": [{"match": {"object_type": "post"}}]}}}' \
  http://localhost:9101/local_all_default/_search | jq '.hits.count'

curl --header "Content-Type: application/json" \
  --request POST -s \
  --data '{"query": {"bool": {"must": [{"match": {"object_type": "comment"}}]}}}' \
  http://localhost:9101/local_all_default/_search | jq '.hits.count'
```
